### ISSUE TITLE
Sematics of object id and name is not defined

### ISSUE THREAD

**Issue by diekhans** [ga4gh/schemas#381]
The semantics of object id and name are not defined or documented.

The reference server assumes that the id fields are under it's control and not portable across server
instances.  

It's undefined how one uniquely identifies objects and object versions across multiple servers. This relates previous discussion of GUIDs.

**Comment by jeromekelleher** [comment]
I guess what we have at the moment is:

- The ``id`` is a server defined opaque string, with no portability requirements at all.
- The ``name`` is a user defined string attached to container objects, which may be null.

Until we've decided what the container objects actually are (see #380), defining GUIDs will be hard.

**Comment by diekhans** [comment]
from @sarahhunt   in
https://github.com/ga4gh/compliance/issues/34#issuecomment-132119410

 In Ensembl our internal ids are database sequence derived, but we did not want to expose them externally as they are not necessarily consistent between releases. When we were setting up the API, there were discussions within GA4GH about adopting data derived ids, so we used placeholders pending the outcome of that discussion. The data set and variant set ids are from simple numeric sequences and the callset and variant ids are the sample/variant names - these are unique within the datasets as we configure them.

**Comment by dglazer** [comment]
I agree with @jeromekelleher 's description of current state. I'd add:
* The `id` must be "id-like" within a given server instance -- every object has exactly one id, and every id maps to exactly one object
* The `name`, since it's user-defined, has no semantic requirements. (Many users may choose to use "id-like" semantics, but servers don't enforce that.)

**Comment by jeromekelleher** [comment]
@dglazer --- agreed, good clarifications.

**Comment by diekhans** [comment]
@dglazer, @jeromekelleher, while I believe this definition of
`id' is the current, undocumented practice, it's not clear if
this is the desired design goal.

The problem stems the goals of GA4GH API not being clearly
defined.  Are we

   a. Defining an API to implement software that is programmed in a similar manner,
      but not necessary interoperable on a data level.  (similar to Java or C standard
      libraries).
or
   b. Defining an API to build an global interoperable, federate network for data
      exchange (similar to what INSDC gives you on the data level).

Either are valid design goals. Without clear definition of what
we are actually building, we are going to have a difficult time
succeeding.

**Comment by dglazer** [comment]
Responding to the two parts of @diekhans comment:

> I believe this definition of `id' is the current, undocumented practice

Good; that means we've now addressed the original goal of this PR, and agree on the current semantics of object id and name. I defer to you on how you want to capture that understanding officially -- maybe a new PR to add a header comment somewhere in schemas/ with the text from this discussion?

>  it's not clear if this is the desired design goal.

* I think your (a), interoperable code, is a prereq for your (b), federated data. 
* we've so far been focused on (a). IMO the current schema at least largely satisfies it; we should continue to prioritize cleaning up any doc or details where that's not true.
* getting to (b) would require agreeing on some sort of GUID for important object types, whether data-based (as we've already done for references) or accession-based. That discussion has started a couple of times, but hasn't had a champion driving for resolution, and the devil is in the details. IMO it's a nice idea, but not yet blocking adoption; if we wanted to push on it, I'd recommend starting with a compelling driver project.

**Comment by diekhans** [comment]
t's been pointed out that the server-local scope of `id' is analogous to SQL foreign keys.  Probably good to explain this is the documentations.
